examples/toy/Ch4/mlir/Dialect.cpp

```diff
@@ -0,0 +1,262 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// =============================================================================
+//
+// This file implements the dialect for the Toy IR: custom type parsing and
+// operation verification.
+//
+//===----------------------------------------------------------------------===//
+
+#include "toy/Dialect.h"
+
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/StandardTypes.h"
+#include "mlir/Transforms/InliningUtils.h"
+#include "include/ShapeInferencePass.h"
+
+using namespace mlir;
+using namespace mlir::toy;
+
+/// This class defines the interface for handling inlining with Toy
+/// operations.
+struct ToyInlinerInterface : public DialectInlinerInterface {
+  using DialectInlinerInterface::DialectInlinerInterface;
+
+  //===--------------------------------------------------------------------===//
+  // Analysis Hooks
+  //===--------------------------------------------------------------------===//
+
+  bool shouldAnalyzeRecursively(Operation *op) const final { return true; }
+
+  /// All operations within toy can be inlined.
+  bool isLegalToInline(Operation *, Region *,
+                       BlockAndValueMapping &) const final {
+    return true;
+  }
+
+  //===--------------------------------------------------------------------===//
+  // Transformation Hooks
+  //===--------------------------------------------------------------------===//
+
+  /// Handle the given inlined terminator by replacing it with a new operation
+  /// as necessary.
+  void handleTerminator(Operation *op,
+                        ArrayRef<Value *> valuesToRepl) const final {
+    // Only "toy.return" needs to be handled here.
+    auto returnOp = cast<ReturnOp>(op);
+
+    // Replace the values directly with the return operands.
+    assert(returnOp.getNumOperands() == valuesToRepl.size());
+    for (const auto &it : llvm::enumerate(returnOp.getOperands()))
+      valuesToRepl[it.index()]->replaceAllUsesWith(it.value());
+  }
+};
+
+/// Override the shape inference interface to add support for toy operations.
+struct ToyShapeInferenceInterface : public DialectShapeInferenceInterface {
+  using DialectShapeInferenceInterface::DialectShapeInferenceInterface;
+
+  virtual bool requiresShapeInference() const final {
+      return true;
+  }
+
+  virtual bool returnsGenericArray(Operation *op) const final {
+      if (op->getNumResults() == 1) {
+        auto arrayTy = op->getResult(0)->getType().cast<RankedTensorType>();
+        return arrayTy.getShape().empty();
+      }
+      return false;
+  }
+
+  void inferShape(Operation *op) {
+      // The add operation is trivial: propagate the input type as is.
+      if (auto addOp = llvm::dyn_cast<AddOp>(op)) {
+        op->getResult(0)->setType(op->getOperand(0)->getType());
```

Comment:
If this is an OpInterface, each of these can be handled directly by the ops themselves.